pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/OpenIDProviderMetadataCache.java

```diff
@@ -0,0 +1,233 @@
+
+    OpenIDProviderMetadataCache(ServiceConfiguration config, AsyncHttpClient httpClient, ApiClient apiClient) {
+        int maxSize = getConfigValueAsInt(config, CACHE_SIZE, CACHE_SIZE_DEFAULT);
+        int refreshAfterWriteSeconds = getConfigValueAsInt(config, CACHE_REFRESH_AFTER_WRITE_SECONDS,
+                CACHE_REFRESH_AFTER_WRITE_SECONDS_DEFAULT);
+        int expireAfterSeconds = getConfigValueAsInt(config, CACHE_EXPIRATION_SECONDS,
+                CACHE_EXPIRATION_SECONDS_DEFAULT);
+        this.httpClient = httpClient;
+        this.wellKnownApi = apiClient != null ? new WellKnownApi(apiClient) : null;
+        AsyncCacheLoader<Optional<String>, OpenIDProviderMetadata> loader = (issuer, executor) -> {
+            if (issuer.isPresent()) {
+                return loadOpenIDProviderMetadataForIssuer(issuer.get());
+            } else {
+                return loadOpenIDProviderMetadataForKubernetesApiServer();
+            }
+        };
+        this.cache = Caffeine.newBuilder()
+                .maximumSize(maxSize)
+                .refreshAfterWrite(refreshAfterWriteSeconds, TimeUnit.SECONDS)
+                .expireAfterWrite(expireAfterSeconds, TimeUnit.SECONDS)
+                .buildAsync(loader);
+    }
+
+    /**
+     * Retrieve the OpenID Provider Metadata for the provided issuer.
+     * <p>
+     * Note: this method does not do any validation on the parameterized issuer. The OpenID Connect discovery
+     * spec requires that the issuer use the HTTPS scheme: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata.
+     * The {@link AuthenticationProviderOpenID} class handles this verification.
+     *
+     * @param issuer - authority from which to retrieve the OpenID Provider Metadata
+     * @return the {@link OpenIDProviderMetadata} for the given issuer. Fail the completable future with
+     * AuthenticationException if any exceptions occur while retrieving the metadata.
+     */
+    CompletableFuture<OpenIDProviderMetadata> getOpenIDProviderMetadataForIssuer(@Nonnull String issuer) {
+        return cache.get(Optional.of(issuer));
+    }
+
+    /**
+     * A loader for the cache that retrieves the metadata from the issuer's /.well-known/openid-configuration endpoint.
+     * @return a connection to the issuer's /.well-known/openid-configuration endpoint. Fails with
+     * AuthenticationException if the URL is malformed or there is an exception while opening the connection
+     */
+    private CompletableFuture<OpenIDProviderMetadata> loadOpenIDProviderMetadataForIssuer(String issuer) {
+        URI uri;
+        try {
+            // TODO URI's normalization follows RFC2396, whereas the spec
+            //  https://openid.net/specs/openid-connect-discovery-1_0.html#NormalizationSteps
+            //  calls for normalization according to RFC3986, which is supposed to obsolete RFC2396
+            uri = URI.create(issuer + "/.well-known/openid-configuration").normalize();
```

Comment:
I've been thinking about this, and I think it makes sense to change it to just drop the `/` when needed. For what it's worth, the call to `normalize()` properly maps `//` to `/`. However, since the spec took the time to spell this out exactly, it makes sense to me to follow it exactly.

Reply:
Note that the OIDC discovery spec indicates this in [Section 4.1](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationRequest):

> If the Issuer value contains a path component, any terminating / MUST be removed before appending /.well-known/openid-configuration. The RP would make the following request to the Issuer https://example.com/issuer1 to obtain its Configuration information, since the Issuer contains a path component:

Reply:
One counterpoint is that we already use this normalization method here in the client:

https://github.com/apache/pulsar/blob/ec102fb024a6ea2b195826778300f20e330dff06/pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/oauth2/protocol/DefaultMetadataResolver.java#L105-L111